torch/_higher_order_ops/associative_scan.py

```diff
@@ -401,9 +443,421 @@ def associative_scan_op_dense(combine_fn, xs, additional_inputs):
+            return g_y * b, g_y * a
+
+        Note: In a real usecase of associative_scan, there may be additional_inputs that participate in the
+        forward as well as in the backward of the scan operator. For the sake of readability those inputs
+        have been omitted in the following example, but are included in the subsequent detailed description below
+
+        The forward output of associative_scan is computed as:
+        ys = associative_scan(combine_fn, xs).
+
+        For example, this computation can be unpacked as:
+        ys_0 = xs_0
+        ys_1 = combine_fn(ys_0, xs_1)
+        ...
+        ys_T = combine_fn(ys_(T-1), xs_T)
+
+        Note: In a real usecase of associative_scan this operation is parallelized from O(T) to O(log(T)).
+
+        Given the ys, the gradients for xs can be computed as follows:
+        We receive the upstream gradients in torch.autograd.Function, i.e., we get g_ys,
+        where g_ys is the vector of all intermediate gradients of the outputs [g_ys_0, g_ys_1, ..., g_ys_T]
+
+        We then proceed to compute the gradients for the xs (g_xs) by computing for every step:
+        the instantaneous for g_x_t, as well as
+        the gradient for the previous intermediate results g_h_t using
+        g_h_t, g_x_t = ctx._combine_fn_bw(ys_(t-1), xs_t)
+
+        For example:
+        g_h_0, g_x_0 = [1, 1]
+        g_h_1, g_x_1 = ctx._combine_fn_bw(ys_0, ys_1)
+        g_h_2, g_x_2 = ctx._combine_fn_bw(xs_0, xs_1)
+        ...
+        g_h_T, g_x_T = ctx._combine_fn_bw(xs_0, xs_1).
+
+        For the example above,
+        g_h = [1, 2, 3, 4]
+        g_x = [1, 1, 2, 6]
+
+        We can then compute the 'gradient transition matrix' h_mat using the instantaneous gradient components g_h_t as
+        h_mat = [[1, g_h_1, g_h_2 . g_h_1, g_h_3 . g_h_2 . g_h_1],
+                 [0, 1    , g_h_2        , g_h_3 . g_h_2        ],
+                 [0, 0    , 1            , g_h_3                ],
+                 [0, 0    , 0            , 1                    ]],
+        which for the example above results in:
+        h_mat = [[1, 2, 6, 24],
+                 [0, 1, 3, 12],
+                 [0, 0, 1,  4],
+                 [0, 0, 0,  1]]
+
+        We then scale the h_mat with the upstream gradient g_ys
+
+        scaled_h_mat = h_mat * g_ys
+        Assuming all 1s for the upstream gradients this would result in:
+        scaled_h_mat = [[1, 2, 6, 24],
+                        [0, 1, 3, 12],
+                        [0, 0, 1,  4],
+                        [0, 0, 0,  1]]
+
+        Sum the h_mat row-wise
+        summed_h_mat = scaled_h_mat.sum(1) # Row-wise summation
+        which would be
+        summed_h_mat = [33, 16, 5, 1]
+
+        and multiply with the instantaneous gradients g_x.
+        g_xs = summed_h_mat * g_x
+
+        g_xs = [33, 16, 5, 1] * [1, 1, 2, 6]
+        g_xs = [33, 16, 10, 6]
+        With this procedure we end up with the
+        the gradients for the xs -> g_xs.
+
```

Comment:
To me this only explains how but doesn't explain why, it's not obvious to me what we're doing here. Can you explain the overall idea behind why this works?